File: sdks/go/transport/command_map.go
package transport

// CommandMapping is the result of mapping a dotted SDK name onto a
// wire-level verb + argument vector.
type CommandMapping struct {
	Command string
	Args    []NexusValue
}

// MapCommand translates an SDK dotted name (`graph.cypher`,
// `db.create`, …) into a wire command plus an argument list. Returns
// nil when the name is not mapped so the caller can fall back to HTTP
// or surface a clear error.
//
// The table must stay in sync with docs/specs/sdk-transport.md §6 and
// with the Rust SDK's sdks/rust/src/transport/command_map.rs.
func MapCommand(dotted string, payload map[string]any) *CommandMapping {
	switch dotted {
	case "graph.cypher":
		q, ok := payload["query"].(string)
		if !ok {
			return nil
		}
		args := []NexusValue{NxStr(q)}
		if params, present := payload["parameters"]; present && params != nil {
			args = append(args, JsonToNexus(params))
		}
		return &CommandMapping{Command: "CYPHER", Args: args}
	case "graph.ping":
		return &CommandMapping{Command: "PING"}
	case "graph.hello":
		return &CommandMapping{Command: "HELLO", Args: []NexusValue{NxInt(1)}}
	case "graph.stats":
		return &CommandMapping{Command: "STATS"}
	case "graph.health":
		return &CommandMapping{Command: "HEALTH"}
	case "graph.quit":
		return &CommandMapping{Command: "QUIT"}
	case "auth.login":
		if key, ok := payload["api_key"].(string); ok && key != "" {
			return &CommandMapping{Command: "AUTH", Args: []NexusValue{NxStr(key)}}
		}
		u, ok1 := payload["username"].(string)
		p, ok2 := payload["password"].(string)
		if !ok1 || !ok2 {
			return nil
		}
		return &CommandMapping{Command: "AUTH", Args: []NexusValue{NxStr(u), NxStr(p)}}

	case "db.list":
		return &CommandMapping{Command: "DB_LIST"}
	case "db.create":
		if n, ok := payload["name"].(string); ok {
			return &CommandMapping{Command: "DB_CREATE", Args: []NexusValue{NxStr(n)}}
		}
		return nil
	case "db.drop":
		if n, ok := payload["name"].(string); ok {
			return &CommandMapping{Command: "DB_DROP", Args: []NexusValue{NxStr(n)}}
		}
		return nil
	case "db.use":
		if n, ok := payload["name"].(string); ok {
			return &CommandMapping{Command: "DB_USE", Args: []NexusValue{NxStr(n)}}
		}
		return nil

	case "schema.labels":
		return &CommandMapping{Command: "LABELS"}
	case "schema.rel_types":
		return &CommandMapping{Command: "REL_TYPES"}
	case "schema.property_keys":
		return &CommandMapping{Command: "PROPERTY_KEYS"}
	case "schema.indexes":
		return &CommandMapping{Command: "INDEXES"}

	case "data.export":
		fmt, ok := payload["format"].(string)
		if !ok {
			return nil
		}
		args := []NexusValue{NxStr(fmt)}
		if q, ok := payload["query"].(string); ok {
			args = append(args, NxStr(q))
		}
		return &CommandMapping{Command: "EXPORT", Args: args}
	case "data.import":
		fmt, ok1 := payload["format"].(string)
		data, ok2 := payload["data"].(string)
		if !ok1 || !ok2 {
			return nil
		}
		return &CommandMapping{Command: "IMPORT", Args: []NexusValue{NxStr(fmt), NxStr(data)}}
	}
	return nil
}

// JsonToNexus — JSON-compatible Go value to NexusValue.
func JsonToNexus(v any) NexusValue {
	switch x := v.(type) {
	case nil:
		return NxNull()
	case bool:
		return NxBool(x)
	case int:
		return NxInt(int64(x))
	case int32:
		return NxInt(int64(x))
	case int64:
		return NxInt(x)
	case float32:
		return NxFloat(float64(x))
	case float64:
		// JSON numbers land here — preserve integer magnitudes.
		if x == float64(int64(x)) {
			return NxInt(int64(x))
		}
		return NxFloat(x)
	case string:
		return NxStr(x)
	case []byte:
		return NxBytes(x)
	case []any:
		out := make([]NexusValue, len(x))
		for i, e := range x {
			out[i] = JsonToNexus(e)
		}
		return NxArray(out)
	case map[string]any:
		pairs := make([]MapEntry, 0, len(x))
		for k, val := range x {
			pairs = append(pairs, MapEntry{Key: NxStr(k), Value: JsonToNexus(val)})
		}
		return NxMap(pairs)
	}
	return NxNull()
}

// NexusToJson — NexusValue to JSON-compatible Go value for user-visible
// surfaces. Keys that are not strings are stringified.
func NexusToJson(v NexusValue) any {
	switch v.Kind {
	case KindNull:
		return nil
	case KindBool, KindInt, KindFloat, KindStr, KindBytes:
		return v.Value
	case KindArray:
		arr := v.Value.([]NexusValue)
		out := make([]any, len(arr))
		for i, e := range arr {
			out[i] = NexusToJson(e)
		}
		return out
	case KindMap:
		pairs := v.Value.([]MapEntry)
		out := make(map[string]any, len(pairs))
		for _, p := range pairs {
			key := ""
			switch p.Key.Kind {
			case KindStr:
				key = p.Key.Value.(string)
			default:
				if s, ok := p.Key.Value.(string); ok {
					key = s
				} else {
					// stringify numbers etc.
					key = ""
				}
			}
			out[key] = NexusToJson(p.Value)
		}
		return out
	}
	return nil
}

File: sdks/php/src/Transport/CommandMap.php
<?php

declare(strict_types=1);

namespace Nexus\SDK\Transport;

/**
 * SDK dotted-name → wire-command mapping. Parity with
 * sdks/rust/src/transport/command_map.rs — 26 entries.
 */
final class CommandMap
{
    /**
     * @param array<string, mixed> $payload
     * @return array{command: string, args: NexusValue[]}|null
     */
    public static function map(string $dotted, array $payload): ?array
    {
        switch ($dotted) {
            case 'graph.cypher':
                if (!isset($payload['query']) || !is_string($payload['query'])) {
                    return null;
                }
                $args = [NexusValue::str($payload['query'])];
                if (array_key_exists('parameters', $payload) && $payload['parameters'] !== null) {
                    $args[] = self::jsonToNexus($payload['parameters']);
                }
                return ['command' => 'CYPHER', 'args' => $args];
            case 'graph.ping':
                return ['command' => 'PING', 'args' => []];
            case 'graph.hello':
                return ['command' => 'HELLO', 'args' => [NexusValue::int(1)]];
            case 'graph.stats':
                return ['command' => 'STATS', 'args' => []];
            case 'graph.health':
                return ['command' => 'HEALTH', 'args' => []];
            case 'graph.quit':
                return ['command' => 'QUIT', 'args' => []];
            case 'auth.login':
                $apiKey = $payload['api_key'] ?? null;
                if (is_string($apiKey) && $apiKey !== '') {
                    return ['command' => 'AUTH', 'args' => [NexusValue::str($apiKey)]];
                }
                $u = $payload['username'] ?? null;
                $p = $payload['password'] ?? null;
                if (!is_string($u) || !is_string($p)) {
                    return null;
                }
                return ['command' => 'AUTH', 'args' => [NexusValue::str($u), NexusValue::str($p)]];
            case 'db.list':
                return ['command' => 'DB_LIST', 'args' => []];
            case 'db.create':
            case 'db.drop':
            case 'db.use':
                $n = $payload['name'] ?? null;
                if (!is_string($n)) {
                    return null;
                }
                $cmd = match ($dotted) {
                    'db.create' => 'DB_CREATE',
                    'db.drop' => 'DB_DROP',
                    'db.use' => 'DB_USE',
                };
                return ['command' => $cmd, 'args' => [NexusValue::str($n)]];
            case 'schema.labels':
                return ['command' => 'LABELS', 'args' => []];
            case 'schema.rel_types':
                return ['command' => 'REL_TYPES', 'args' => []];
            case 'schema.property_keys':
                return ['command' => 'PROPERTY_KEYS', 'args' => []];
            case 'schema.indexes':
                return ['command' => 'INDEXES', 'args' => []];
            case 'data.export':
                $fmt = $payload['format'] ?? null;
                if (!is_string($fmt)) {
                    return null;
                }
                $args = [NexusValue::str($fmt)];
                $q = $payload['query'] ?? null;
                if (is_string($q)) {
                    $args[] = NexusValue::str($q);
                }
                return ['command' => 'EXPORT', 'args' => $args];
            case 'data.import':
                $fmt = $payload['format'] ?? null;
                $d = $payload['data'] ?? null;
                if (!is_string($fmt) || !is_string($d)) {
                    return null;
                }
                return ['command' => 'IMPORT', 'args' => [NexusValue::str($fmt), NexusValue::str($d)]];
        }
        return null;
    }

    public static function jsonToNexus(mixed $v): NexusValue
    {
        if ($v === null) {
            return NexusValue::null();
        }
        if (is_bool($v)) {
            return NexusValue::bool($v);
        }
        if (is_int($v)) {
            return NexusValue::int($v);
        }
        if (is_float($v)) {
            return NexusValue::float($v);
        }
        if (is_string($v)) {
            return NexusValue::str($v);
        }
        if (!is_array($v)) {
            return NexusValue::null();
        }
        // List vs assoc.
        if (array_keys($v) === range(0, count($v) - 1)) {
            return NexusValue::array(array_map(
                fn ($e) => self::jsonToNexus($e),
                $v,
            ));
        }
        $pairs = [];
        foreach ($v as $k => $val) {
            $pairs[] = [NexusValue::str((string) $k), self::jsonToNexus($val)];
        }
        return NexusValue::map($pairs);
    }

    public static function nexusToJson(NexusValue $v): mixed
    {
        return match ($v->kind) {
            NexusValueKind::Null => null,
            NexusValueKind::Bool,
            NexusValueKind::Int,
            NexusValueKind::Float,
            NexusValueKind::Str,
            NexusValueKind::Bytes => $v->value,
            NexusValueKind::Array => array_map(
                fn (NexusValue $e) => self::nexusToJson($e),
                $v->value,
            ),
            NexusValueKind::Map => self::mapToAssoc($v->value),
        };
    }

    /**
     * @param array<int, array{0: NexusValue, 1: NexusValue}> $pairs
     * @return array<string, mixed>
     */
    private static function mapToAssoc(array $pairs): array
    {
        $out = [];
        foreach ($pairs as $pair) {
            $key = $pair[0]->kind === NexusValueKind::Str
                ? (string) $pair[0]->value
                : (string) $pair[0]->value;
            $out[$key] = self::nexusToJson($pair[1]);
        }
        return $out;
    }
}
